### DISCUSSION TITLE
Publisher effects resulting in long-running actions?

### DISCUSSION THREAD

**Discussion by krze** [pointfreeco/swift-composable-architecture#3531] [Q&A]
I found the super-handy `signpost()` function on reducers and decided to take a look in instruments to check performance. I found something I don't understand, maybe I'm interpreting it wrong:

In my app, I have a Settings view that you can enter from the main view. When the Settings view appears, I send an `onAppear` Action to its reducer. In response to the `onAppear` Action, that reducer sets up initial state and returns a `publisher` effect that maps another publisher to fire another `Action` when that publish receives a new value.

While viewing the signposts in instruments for my Settings reducer's `onAppear` Action, I see that it keeps running even after the view is dismissed. While this is understandable as I use `scope` to create the Settings reducer from the app's main reducer, I don't understand why the `Action` itself is long-running. Here's a screenshot from instruments, I apologize for cutting off some of the entry due to propriety but bear with me:
<img width="982" alt="image" src="https://github.com/user-attachments/assets/bab0783c-f936-446d-9c75-220a11be047a" />

In this instruments run, I entered and left the Settings view twice. After the initial `onAppear` the `Action` kept running according to the signposts. This is under the `Action` signpost category, not the `Effect` category

If I send `onAppear` to the store using `await store.send(.onAppear).finish()` this stops occurring.

Should I be calling `finish()` on sending actions with long-running publishers? Or is instruments potentially not reporting this correctly? Or, alternatively, is this a non-issue?

**Comment by mbrandonw** [comment]
Hi @krze, hypo will need to provide more information of how you are presenting the settings screen. Can you show the reducer code that executed when presenting settings, as well as the view code that presents the settings view?

**Reply by krze** [reply]
Hi @mbrandonw 

It's been a while and the code as changed so that the settings screen is no longer using a Publisher. However I can share what the setup _was_ when the issue occured.

Here's a snippet of the `SettingsView`'s body. The `store` is a simple `let store: StoreOf<SettingsReducer>` 

```
    public var body: some View {
        WithPerceptionTracking {
            Group {
                if platform.isTvOS {
                    SettingsTvOSView(store: store, pageFocusState: $pageFocusState)
                } else {
                    SettingsIOSView(store: store)
                }
            }
            .onAppear {
               // This was the setup when the above behavior still occurred. It was since changed to use `.finish()`
               store.send(.onAppear) 
            }
        }
    }
```

Here's what was going on in the `SettingsReducer`

```
    private var settingsReducerBody: some Reducer<State, Action> {
        Reduce { state, action in
            switch action {
            case .onAppear:
                // Removed a number of lines that were setting up state
                return .publisher {
                    overlayCoordinator.events.map {
                        SettingsReducer.Action.overlayPresentation(.receiveOverlayEvent($0))
                    }
                }
         // Snipped the rest of the reducer body
```
